packages/fleet/lib/supervisor.ts

```diff
@@ -0,0 +1,279 @@
+import type { DatabaseClient } from './db/client.js';
+import { logger } from './utils/logger.js';
+import * as nodes from './models/nodes.js';
+import * as deployments from './models/deployments.js';
+import { Err, Ok } from '@nangohq/utils';
+import type { Result } from '@nangohq/utils';
+import { FleetError } from './utils/errors.js';
+import type { Deployment, Node } from './types.js';
+import { setTimeout } from 'node:timers/promises';
+import type { NodeProvider } from './node-providers/node_provider.js';
+
+type Operation =
+    | { type: 'CREATE'; routingId: Node['routingId']; deployment: Deployment }
+    | { type: 'START'; node: Node }
+    | { type: 'FAIL'; node: Node; reason: 'starting_timeout_reached' }
+    | { type: 'OUTDATE'; node: Node }
+    | { type: 'TERMINATE'; node: Node }
+    | { type: 'REMOVE'; node: Node }
+    | { type: 'WARN'; node: Node };
+
+type SupervisorState = 'stopped' | 'running' | 'stopping';
+
+export const STATE_TIMEOUT_MS = {
+    STARTING: 5 * 60 * 1000,
+    FINISHING: 24 * 60 * 60 * 1000,
+    TERMINATED: 7 * 24 * 60 * 60 * 1000,
+    ERROR: 7 * 24 * 60 * 60 * 1000
+};
+
+export class Supervisor {
+    private state: SupervisorState = 'stopped';
+    private dbClient: DatabaseClient;
+    private nodeProvider: NodeProvider;
+
+    constructor({ dbClient, nodeProvider }: { dbClient: DatabaseClient; nodeProvider: NodeProvider }) {
+        this.dbClient = dbClient;
+        this.nodeProvider = nodeProvider;
+    }
+
+    public async start(): Promise<void> {
+        if (this.state === 'running') {
+            logger.warn('Supervisor is already running');
+            return;
+        }
+
+        this.state = 'running';
+        await this.loop();
+    }
+
+    public async stop(): Promise<void> {
+        this.state = 'stopping';
+        logger.info('Supervisor stopped');
+
+        // wait for the loop to finish or timeout
+        const waitForStopped = async () => {
+            while (this.state !== 'stopped') {
+                await setTimeout(100);
+            }
+        };
+        await Promise.race([waitForStopped(), setTimeout(60000)]);
+    }
+
+    public async tick(): Promise<void> {
+        // TODO: trace
+        try {
+            const plan = await this.plan();
+            if (plan.isOk()) {
+                await this.executePlan(plan.value);
+            } else {
+                logger.error('Supervision error:', plan.error);
+                await setTimeout(1000);
+            }
+        } catch (error) {
+            logger.error('Supervision error:', error);
+        }
+    }
+
+    private async loop(): Promise<void> {
+        while (this.state === 'running') {
+            // TODO: Lock to prevent multiple instances from running
+            await this.tick();
+        }
+        this.state = 'stopped';
+    }
+
+    private async plan(cursor?: number): Promise<Result<Operation[]>> {
```

Comment:
Would it be easier going forward for operations to be an object with a type and a `() => void` function attached, and then executing the plan just loops and calls the functions?

Reply:
why no. The advantage of having values instead of opaque functions is that you can optimize the plan before to execute it. I don't think Render API can do it and even if it does I am not planning to support it but we could for instance takes all the CREATE operations and combine them in a unique call to the node provider.